partiql-ast/src/main/java/org/partiql/ast/IntervalQualifier.java

```diff
@@ -0,0 +1,272 @@
+         * @return the start field of the range
+         */
+        @NotNull
+        public DatetimeField getStartField() {
+            return this.startField;
+        }
+
+        /**
+         * Returns the precision of the leading field.
+         * @return the precision of the leading field
+         */
+        @Nullable
+        public Integer getStartFieldPrecision() {
+            return this.leadingFieldPrecision;
+        }
+
+        /**
+         * Returns the end field of the range.
+         * @return the end field of the range
+         */
+        @NotNull
+        public DatetimeField getEndField() {
+            return this.endField;
+        }
+
+        /**
+         * Returns the fractional precision of the end field. This may only be present if the end
+         * field is {@link DatetimeField#SECOND}.
+         * @return the fractional precision of the end field
+         */
+        @Nullable
+        public Integer getEndFieldFractionalPrecision() {
+            return this.endFieldFractionalPrecision;
+        }
+
+        /**
+         * Constructs a new Range instance.
+         * @param startField the start field of the range
+         * @param startFieldPrecision the precision of the leading field
+         * @param endField the end field of the range
+         * @param endFieldFractionalPrecision the fractional precision of the end field
+         */
+        public Range(
+                @NotNull DatetimeField startField,
+                @Nullable Integer startFieldPrecision,
+                @NotNull DatetimeField endField,
+                @Nullable Integer endFieldFractionalPrecision
+        ) throws IllegalArgumentException {
+            checkStartField(startField, startFieldPrecision);
+            checkEndField(endField, endFieldFractionalPrecision);
+            this.startField = startField;
+            this.leadingFieldPrecision = startFieldPrecision;
+            this.endField = endField;
+            this.endFieldFractionalPrecision = endFieldFractionalPrecision;
+        }
+
+        /**
+         * Checks that the start field utilizes a valid datetime field and that the leading field precision is valid.
+         * @param startField the start field of the range
+         * @param startFieldPrecision the precision of the leading field
+         * @throws IllegalArgumentException when the start field is not a valid datetime field or the leading field precision is not valid
+         */
+        private static void checkStartField(
+                @NotNull DatetimeField startField,
+                @Nullable Integer startFieldPrecision
+        ) throws IllegalArgumentException {
+            int startCode = startField.code();
+            if (!ALLOWABLE_START_FIELDS.contains(startCode)) {
+                throw new IllegalArgumentException("Invalid interval start field: " + startField.name());
+            }
+            if (startFieldPrecision != null && startFieldPrecision <= 0) {
```

Comment:
The spec says the minimum here is `0` and the maximum here is implementation defined, but I would presume there is a useful maximum we could put in place?